numba/tests/test_typedlist.py

```diff
@@ -101,6 +101,55 @@ def test_basic(self):
         # index
         self.assertEqual(l.index(15), 4)
 
+    def test_unsigned_access(self):
+        L = List.empty_list(int32)
+        ui32_0 = types.uint32(0)
+        ui32_1 = types.uint32(1)
+        ui32_2 = types.uint32(2)
+
+        # insert
+        L.append(types.uint32(10))
+        L.append(types.uint32(11))
+        L.append(types.uint32(12))
+        self.assertEqual(len(L), 3)
+
+        # getitem
+        self.assertEqual(L[ui32_0], 10)
+        self.assertEqual(L[ui32_1], 11)
+        self.assertEqual(L[ui32_2], 12)
+
+        # setitem
+        L[ui32_0] = 123
+        L[ui32_1] = 456
+        L[ui32_2] = 789
+        self.assertEqual(L[ui32_0], 123)
+        self.assertEqual(L[ui32_1], 456)
+        self.assertEqual(L[ui32_2], 789)
+
+        # index
+        ui32_123 = types.uint32(123)
+        ui32_456 = types.uint32(456)
+        ui32_789 = types.uint32(789)
+        self.assertEqual(L.index(ui32_123), 0)
+        self.assertEqual(L.index(ui32_456), 1)
+        self.assertEqual(L.index(ui32_789), 2)
+
+        # delitem
+        L.__delitem__(ui32_2)
+        del L[ui32_1]
+        self.assertEqual(len(L), 1)
+        self.assertEqual(L[ui32_0], 123)
+
+        # pop
+        L.append(2)
+        L.append(3)
+        L.append(4)
+        self.assertEqual(len(L), 4)
+        self.assertEqual(L.pop(), 4)
+        self.assertEqual(L.pop(ui32_2), 3)
+        self.assertEqual(L.pop(ui32_1), 2)
+        self.assertEqual(L.pop(ui32_0), 123)
+
```

Comment:
I think it'd be good to include a test here to confirm that the originally reported issue is indeed fixed (that `prange()` interaction is broken without this patch).

Reply:
unfortunately, it's breaking with the following code:
```python
@njit
def foo(a):
    r = types.uint64(3)
    for i in prange(r):
        print(r[i])

a = typed.List.empty_list(types.uint64)
a.append(types.uint64(1))
a.append(types.uint64(1))

```

Reply:
The `r[i]` is indexing a scalar variable in the above. This is a fixed-up reproducer:
```python
from numba import njit, types, typed, prange

@njit
def foo(a):
    r = types.uint64(3)
    for i in prange(r):
        print(a[i])

a = typed.List.empty_list(types.uint64)
a.append(types.uint64(10))
a.append(types.uint64(20))
a.append(types.uint64(30))
foo(a)
```
note `prange` interprets as `range` in a non-parallel target.

Suspect the problem is here:
https://github.com/numba/numba/blob/a58191d5fba0f1deb3f8a3532facaf836ebecb63/numba/listobject.py#L455-L467

where the `operator.add` with a signed int and an unsigned int as arguments (`len(l) + index `) would get a return type of a float, this is a NumPy casting rule. Expect this would fix:
```python
@register_jitable
def handle_index(l, index):
    if index < 0:
        index = type(index)(len(l)) + index
    # check that the index is in range
    if not (0 <= index < len(l)):
        raise IndexError("list index out of range")
    return index
```